### PR TITLE
Сенина - Лабораторная работа #1

### PR DESCRIPTION
**Pull request by AnastasiaaSenina**
@JuliaKamelina @UsovaMA please review

### CODE REVIEW

File: code/float_vector_metrics/model/test_vector_metrics.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        b = [1.2, 1.4, 1.8]
+        metrics_calculator = VectorMetrics()
+        calc_value = metrics_calculator.l4(a, b)
+        self.assertAlmostEqual(calc_value, 0.4756828)
+
+    def test_can_calculate_linf_metric(self):
+        a = [1.2, 1.5, 1.8]
+        b = [3.3, 8.7, 1.7]
+        metrics_calculator = VectorMetrics()
+        calc_value = metrics_calculator.l4(a, b)
```
Comment (JuliaKamelina) [inline code/float_vector_metrics/model/test_vector_metrics.py:44]:
```suggestion
        calc_value = metrics_calculator.linf(a, b)
```

File: code/float_vector_metrics/model/test_vector_metrics.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        metrics_calculator = VectorMetrics()
+        with self.assertRaises(TypeError):
+            metrics_calculator.l4(a, b)
+
+    def test_can_raise_error_when_vector_of_equal_length(self):
+        a = [1.2, 1.5, 1.8]
+        b = [5]
+        metrics_calculator = VectorMetrics()
+        with self.assertRaises(ValueError):
+            metrics_calculator.l3(a, b)
```
Comment (JuliaKamelina) [inline code/float_vector_metrics/model/test_vector_metrics.py:59]: Let's add more tests for invalid argument like `b = ['1.2', '1.5', '1.8']`

File: code/float_vector_metrics/model/test_vector_metrics.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        self.assertAlmostEqual(calc_value, 7.2129911)
+
+    def test_can_raise_error_when_vector_not_list(self):
+        a = [1.2, 1.5, 1.8]
+        b = 5
+        metrics_calculator = VectorMetrics()
+        with self.assertRaises(TypeError):
+            metrics_calculator.l4(a, b)
+
+    def test_can_raise_error_when_vector_of_equal_length(self):
```
Comment (JuliaKamelina) [inline code/float_vector_metrics/model/test_vector_metrics.py:54]:
```suggestion
    def test_can_raise_error_when_vector_of_unequal_length(self):
```

File: code/float_vector_metrics/model/float_vector_metrics.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+    @check_arg(List, List)
+    def linf(self, vector_a: List, vector_b: List):
+
+        return max(abs(a - b) for a, b in zip(vector_a, vector_b))
+
+    @staticmethod
+    def __lp(vector_a: List, vector_b: List, p: int):
+
+        return sum(abs(a - b)**p for a, b in zip(vector_a, vector_b))**(1 / p)
```
Comment (UsovaMA) [inline code/float_vector_metrics/model/float_vector_metrics.py:50]: Add a parameter p validation check (p>0) and test on it

File: code/float_vector_metrics/model/test_vector_metrics.py (hunk 2 above)
Comment (AnastasiaaSenina) [reply code/float_vector_metrics/model/test_vector_metrics.py:59]: Done.

File: code/float_vector_metrics/model/float_vector_metrics.py (hunk 4 above)
Comment (AnastasiaaSenina) [reply code/float_vector_metrics/model/float_vector_metrics.py]: Done.

**Comment by UsovaMA** [comment]
:+1:

**Review by JuliaKamelina — APPROVED** [review]
👍

**Comment by AnastasiaaSenina** [comment]
@kirill-korniakov

**Review by sawirani — APPROVED** [review]

**Pull request merged**